Change temporary entrypoints to be lazily allocated

davidwrighton opened the pull request:
This change moves temporary entrypoints to be allocated when they are needed instead of eagerly at type load time

As you can see, with a very small test case the wins are significant, and notably, reducing allocations in the stub heaps is nice as they are difficult to allocate efficiently as they require a bunch of memory mapping operations. In addition, with this change I'm removing the concept of Compact Entrypoints, as they are no longer significantly beneficial as of the advent of tiered compilation, and would be difficult to integrate with this change.

```
Baseline
Loader Heap:
----------------------------------------
LowFrequencyHeap:     Size: 0xf0000 (983040) bytes total.
HighFrequencyHeap:    Size: 0x16a000 (1482752) bytes total, 0x3000 (12288) bytes wasted.
StubHeap:             Size: 0x1000 (4096) bytes total.
FixupPrecodeHeap:     Size: 0x168000 (1474560) bytes total.
NewStubPrecodeHeap:   Size: 0x18000 (98304) bytes total.
IndirectionCellHeap:  Size: 0x1000 (4096) bytes total.
CacheEntryHeap:       Size: 0x1000 (4096) bytes total.
Total size:           Size: 0x3dd000 (4050944) bytes total, 0x3000 (12288) bytes wasted.


PR
Loader Heap:
----------------------------------------
LowFrequencyHeap:     Size: 0xef000 (978944) bytes total.
HighFrequencyHeap:    Size: 0x1ad000 (1757184) bytes total, 0x3000 (12288) bytes wasted.
StubHeap:             Size: 0x1000 (4096) bytes total.
FixupPrecodeHeap:     Size: 0x38000 (229376) bytes total.
NewStubPrecodeHeap:   Size: 0x8000 (32768) bytes total.
IndirectionCellHeap:  Size: 0x1000 (4096) bytes total.
CacheEntryHeap:       Size: 0x1000 (4096) bytes total.
Total size:           Size: 0x2df000 (3010560) bytes total, 0x3000 (12288) bytes wasted.
```

What can be seen here, is that the size of the Precode heaps shrinks dramatically, and the size of the HighFrequency heap grows a little bit. With a larger test case (launching powershell with a `-c exit` command line argument,  the gains are still very similar. Also with powershell, I have a simple benchmark which runs `pwsh.exe -c exit` 100 times and measures the total execution time. The performance improvement is around 1% which I consider to be quite a nice result. 

| Architecture/Version | Total Size | Process Execution Time |
| ----- | --- | --- |
| Baseline X86 | 9,420,800 bytes | 23.33 seconds |
| PR X86 | 6,348,800 bytes | 23.18 seconds |
| Baseline X64 | 13,586,432 bytes | 23.23 seconds |
| PR X64 | 10,440,704 bytes | 23.15 seconds |

davidwrighton committed:
WorkingOnIt

davidwrighton committed:
It basically works for a single example.

Baseline
Loader Heap:
----------------------------------------
System Domain:        7ffab916ec00
LoaderAllocator:      7ffab916ec00
LowFrequencyHeap:     Size: 0xf0000 (983040) bytes total.
HighFrequencyHeap:    Size: 0x16a000 (1482752) bytes total, 0x3000 (12288) bytes wasted.
StubHeap:             Size: 0x1000 (4096) bytes total.
FixupPrecodeHeap:     Size: 0x168000 (1474560) bytes total.
NewStubPrecodeHeap:   Size: 0x18000 (98304) bytes total.
IndirectionCellHeap:  Size: 0x1000 (4096) bytes total.
CacheEntryHeap:       Size: 0x1000 (4096) bytes total.
Total size:           Size: 0x3dd000 (4050944) bytes total, 0x3000 (12288) bytes wasted.

Compare
Loader Heap:
----------------------------------------
System Domain:        7ff9eb49dc00
LoaderAllocator:      7ff9eb49dc00
LowFrequencyHeap:     Size: 0xef000 (978944) bytes total.
HighFrequencyHeap:    Size: 0x1b2000 (1777664) bytes total, 0x3000 (12288) bytes wasted.
StubHeap:             Size: 0x1000 (4096) bytes total.
FixupPrecodeHeap:     Size: 0x70000 (458752) bytes total.
NewStubPrecodeHeap:   Size: 0x10000 (65536) bytes total.
IndirectionCellHeap:  Size: 0x1000 (4096) bytes total.
CacheEntryHeap:       Size: 0x1000 (4096) bytes total.
Total size:           Size: 0x324000 (3293184) bytes total, 0x3000 (12288) bytes wasted.

LowFrequencyHeap is 4KB bigger
HighFrequencyHeap is 288KB bigger
FixupPrecodeHeap is 992KB smaller
NewstubPrecodeHeap is 32KB smaller

davidwrighton committed:
If there isn't a parent methodtable and the slot matches... then it by definition the method is defining the slot

davidwrighton committed:
Fix a couple more issues found when running a subset of the coreclr tests

davidwrighton committed:
Get X86 building again

davidwrighton committed:
Attempt to use a consistent api to force slots to be set

davidwrighton committed:
Put cache around RequiresStableEntryPoint

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into change_temporary_entrypoints

davidwrighton committed:
Fix typo

davidwrighton committed:
Fix interop identified issue where we sometime set a non Precode into an interface

davidwrighton committed:
Move ARM and X86 to disable compact entry points

davidwrighton committed:
Attempt to fix build breaks

davidwrighton committed:
fix typo

davidwrighton committed:
Fix another Musl validation issue

davidwrighton committed:
More tweaks around NULL handling

davidwrighton committed:
Hopefully the last NULL issue

davidwrighton committed:
Fix more NULL issues

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into change_temporary_entrypoints

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into change_temporary_entrypoints

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into change_temporary_entrypoints

davidwrighton committed:
Fixup obvious issues

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into change_temporary_entrypoints

davidwrighton committed:
Fix allocation behavior so we don't free the data too early or too late

davidwrighton committed:
Fix musl validation issue

davidwrighton committed:
Fix tiered compilation

davidwrighton committed:
Remove Compact Entrypoint logic

AaronRobinsonMSFT commented on src/coreclr/vm/method.hpp:
```diff
@@ -162,8 +162,7 @@ enum MethodDescFlags
     PTR_MethodDescVersioningState VersioningState;
-
-    // [TODO] Move temporary entry points here.
+    PCODE m_pTemporaryEntryPoint;
```
```suggestion
    PCODE TemporaryEntryPoint;
```

This is a public field and I'd prefer to avoid this sort of hungarian notion.

AaronRobinsonMSFT commented on src/coreclr/debug/daccess/request.cpp:
```diff
@@ -435,7 +435,7 @@ ClrDataAccess::GetMethodTableSlot(CLRDATA_ADDRESS mt, unsigned int slot, CLRDATA
             if (pMD->GetSlot() == slot)
             {
-                *value = pMD->GetMethodEntryPoint();
+                *value = pMD->GetMethodEntryPoint_NoAlloc();
```
Is `NULL` a valid return here?

jkotas replied:
Returning NULL is going to break the method enumeration done by DumpMT. Related discussion: https://github.com/dotnet/diagnostics/pull/4760#discussion_r1653807525 . We may want to introduce a better method enumeration DAC API.

AaronRobinsonMSFT commented on src/coreclr/vm/arm/stubs.cpp:
```diff
@@ -1381,6 +1381,7 @@ VOID StubLinkerCPU::EmitShuffleThunk(ShuffleEntry *pShuffleEntryArray)
 
 void StubLinkerCPU::ThumbEmitTailCallManagedMethod(MethodDesc *pMD)
 {
+    pMD->EnsureSlotFilled();
```
This looks like an inoccuous little function. Seems appropriate to add a comment and the missing contracts.

AaronRobinsonMSFT commented on src/coreclr/vm/arm64/stubs.cpp:
```diff
@@ -1625,6 +1625,8 @@ void StubLinkerCPU::EmitCallLabel(CodeLabel *target, BOOL fTailCall, BOOL fIndir
 
 void StubLinkerCPU::EmitCallManagedMethod(MethodDesc *pMD, BOOL fTailCall)
 {
+    pMD->EnsureTemporaryEntryPoint(pMD->GetLoaderAllocator());
```
Similar to the `arm/stubs.cpp` comment. A comment and contract seems appropriate here.

AaronRobinsonMSFT commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -1552,7 +1552,7 @@ extern "C" void QCALLTYPE Interlocked_MemoryBarrierProcessWide()
 {
     CONTRACTL{
-        NOTHROW;
+        THROWS;
```
What has changed that causes this to throw?

jkotas replied:
This calls GetRestoredSlot that THROWS after this change.

AaronRobinsonMSFT commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -1809,7 +1809,7 @@ static WORD g_slotBeginWrite, g_slotEndWrite;
 {
     CONTRACTL{
-        NOTHROW;
+        THROWS;
```
Do these really need to throw now?

jkotas replied:
Calls GetRestoredSlot

AaronRobinsonMSFT commented on src/coreclr/vm/i386/stublinkerx86.cpp:
```diff
@@ -3060,6 +3060,8 @@ VOID StubLinkerCPU::EmitComputedInstantiatingMethodStub(MethodDesc* pSharedMD, s
 #ifdef TARGET_AMD64
 VOID StubLinkerCPU::EmitLoadMethodAddressIntoAX(MethodDesc *pMD)
 {
+    pMD->EnsureSlotFilled();
```
All these should have comments. I'll stop commenting.

jkotas replied:
Instead of comments, it may be better to introduce a method that fill the slot and returns its value or address, and call them instead of `GetStableEntryPoint` or `GetAddrOfSlot` below. It would make it self-describing.

davidwrighton replied:
It turns out we already had that function, its TryGetMulticallableAddrOfCode. The behavior is almost what we want, so I've added a new flag, and now this path generates very slightly more optimal code. (TryGetMulticallableAddrOfCode has a slight optimization around handling of Tier1 code.)

AaronRobinsonMSFT commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -5714,7 +5714,7 @@ HCIMPL3(void, JIT_VTableProfile32, Object* obj, CORINFO_METHOD_HANDLE baseMethod
     _ASSERTE(slot < pBaseMD->GetMethodTable()->GetNumVirtuals());
 
-    MethodDesc* pMD = pMT->GetMethodDescForSlot(slot);
+    MethodDesc* pMD = pMT->GetMethodDescForSlot_NoThrow(slot);
```
Does this mean it can return `NULL`? If so, what is protecting that below?

davidwrighton replied:
No, it returns what it would have before, but its a slightly different algorithm that doesn't use any allocation. Its a bit slower, but perf results indicates that it is ok.

AaronRobinsonMSFT commented on src/coreclr/vm/frames.cpp:
```diff
@@ -612,7 +612,7 @@ MethodDesc* StubDispatchFrame::GetFunction()
         if (m_pRepresentativeMT != NULL)
         {
-            pMD = m_pRepresentativeMT->GetMethodDescForSlot(m_representativeSlot);
+            pMD = m_pRepresentativeMT->GetMethodDescForSlot_NoThrow(m_representativeSlot);
```
Is `NULL` okay here?

davidwrighton replied:
No, but that's ok since this function doesn't have that behavior. Its just a different algorithm for producing the same result.

AaronRobinsonMSFT replied:
My point is if this is not supposed to return a `NULL` and we're calling the "NoThrow" then at the very least we need an assert to convey that assumption.

AaronRobinsonMSFT commented on src/coreclr/vm/method.inl:
```diff
@@ -6,6 +6,30 @@
+{
+    LIMITED_METHOD_CONTRACT;
+
+    // Just like above, but without the assert. This is used in the path which initializes the flag.
```
Why does this matter? It seems very odd to have two different functions that do the same thing except for an debug only assert.

AaronRobinsonMSFT replied:
At a minimum the assert one should call the one without the assert.

AaronRobinsonMSFT commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -7765,11 +7752,23 @@ Module *MethodTable::GetDefiningModuleForOpenType()
     RETURN NULL;
 }
 
+PCODE MethodTable::GetRestoredSlotIfExists(DWORD slotNumber)
```
Is the "If exists" concept handled in the `GetSlot()`? Looking at this implementation the name is very unsettling.

AaronRobinsonMSFT replied:
Based on my reading of the existing code the only difference here is an assert. We should not be creating extra functions with guarantees that are enforced only by asserts.

AaronRobinsonMSFT commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -7780,7 +7779,22 @@ PCODE MethodTable::GetRestoredSlot(DWORD slotNumber)
+    {
+        // This is a slot that has not been filled in yet. This can happen if we are
+        // looking at a slot which has not yet been given a temporary entry point.
+        MethodDesc *pMD = GetCanonicalMethodTable()->GetMethodDescForSlot_NoThrow(slotNumber);
```
What is returned if it would throw?

davidwrighton replied:
The same value. The difference is that the throwing variant works by filling in the slot and then computing the `MethodDesc` from that, and the NoThrow variant does a lookup for the MethodDesc but does not fill in the slot with a code pointer. I've added a comment to the NoThrow variant to this effect.

jkotas commented:
Could you please update https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/method-descriptor.md to match the current state? For example, the first paragraph in https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/method-descriptor.md#method-slots is invalidated by this change.

davidwrighton committed:
Add new ISOSDacInterface15 api

davidwrighton committed:
Fix some naming of NoAlloc to a more clear IfExists suffix

davidwrighton committed:
Remove way in which GetTemporaryEntryPoint behaves differently for DAC builds, and then remove GetTemporaryEntrypoint usage from DAC entirely in favor of GetTemporaryEntryPointIfExists

davidwrighton committed:
Attempt to reduce most of the use of EnsureSlotFilled. Untested, but its late.

davidwrighton committed:
Fix the build before sending to github

jkotas commented on src/coreclr/inc/sospriv.idl:
```diff
@@ -519,3 +519,13 @@ interface ISOSDacInterface14 : IUnknown
+]
+interface ISOSDacInterface15 : IUnknown
+{
+    HRESULT GetMethodTableSlotMethodDesc(CLRDATA_ADDRESS mt, unsigned int slot, CLRDATA_ADDRESS *value);
```
Sharing comment from offline conversation:

We may want to make the new API return (optional) code pointer, (optional) MethodDesc, and optional (Module,methodDef) pair. We should always have at least one of those. The caller can then figure out the details based on what's available.

This would make the API work in case we materialize MethodDescs lazily.

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into change_temporary_entrypoints

davidwrighton committed:
Fix unix build break, and invalid assert

davidwrighton committed:
Improve assertion checks to validate that we don't allocate temporary entrypoints that will be orphaned if the type doesn't actually end up published.

davidwrighton committed:
Remove unused parameters and add contracts

davidwrighton committed:
Update method-descriptor.md

davidwrighton committed:
Fix musl validation issue

davidwrighton committed:
Adjust SOS api to be an enumerator

jkotas commented on src/coreclr/debug/daccess/daccess.cpp:
```diff
@@ -8340,6 +8340,44 @@ HRESULT DacMemoryEnumerator::Next(unsigned int count, SOSMemoryRegion regions[],
+    if (!pCount)
+        return E_POINTER;
+
+    mMethods.GetCount();
```
```suggestion
    *pCount = mMethods.GetCount();
```

jkotas commented on docs/design/coreclr/botr/method-descriptor.md:
```diff
@@ -85,7 +85,9 @@ DWORD MethodDesc::GetAttrs()
 ------------
 
-Each MethodDesc has a slot, which contains the entry point of the method. The slot and entry point must exist for all methods, even the ones that never run like abstract methods. There are multiple places in the runtime that depend on the 1:1 mapping between entry points and MethodDescs, making this relationship an invariant.
+Each MethodDesc has a slot, which contains the entry point of the method. The slot must exist for all methods, even the ones that never run like abstract methods. There are multiple places in the runtime that depend on the 1:1 mapping between entry points and MethodDescs, making this relationship an invariant.
```
```suggestion
Each MethodDesc has a slot, which contains the current entry point of the method. The slot must exist for all methods, even the ones that never run like abstract methods. There are multiple places in the runtime that depend on mapping between entry points and MethodDescs.
```

jkotas commented on src/coreclr/vm/arm/stubs.cpp:
```diff
@@ -1381,14 +1381,18 @@ VOID StubLinkerCPU::EmitShuffleThunk(ShuffleEntry *pShuffleEntryArray)
     }
     else
     {
+        _ASSERTE(!pMD->HasStableEntryPoint());
```
Can this assert fire due to race conditions?

davidwrighton replied:
Oh, hmm...  yeah, it probably could. Sigh. I'd like to see the results of a full PR test pass before I get rid of these asserts though.

jkotas commented on src/coreclr/vm/methodtable.h:
```diff
@@ -1621,9 +1639,14 @@ class MethodTable
 
     MethodDesc* GetMethodDescForSlot(DWORD slot);
 
+    // This api produces the same result as GetMethodDescForSlot_NoThrow, but it uses a variation on the
```
```suggestion
    // This api produces the same result as GetMethodDescForSlot_NoThrow, but it uses a variation on the
```
```suggestion
    // This api produces the same result as GetMethodDescForSlot, but it uses a variation on the
```
?

jkotas commented on src/coreclr/vm/methodtable.inl:
```diff
@@ -415,6 +415,49 @@ inline MethodDesc* MethodTable::GetMethodDescForSlot(DWORD slot)
+    }
+
+    // This is an optimization that we can take advantage of if we're trying to get the MethodDesc
+    // for an interface virtual, since their slots usually point to stub.
```
```suggestion
    // for an interface virtual, since their slots usually point to stub.
```
```suggestion
    // for an interface virtual, since their slots usually point to stub.
```
```suggestion
    // for an interface virtual, since their slots point to stub.
```

jkotas replied:
```suggestion
    // for an interface virtual, since their slots usually point to stub.
```
```suggestion
    // for an interface virtual, since their slots point to stub.
```

davidwrighton committed:
Fix assertion issues noted
Fix ISOSDacInterface15 to actually work

davidwrighton committed:
Remove GetRestoredSlotIfExists
- Its the same as GetSlot ....  just replace it with that function.

davidwrighton committed:
Update src/coreclr/debug/daccess/daccess.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

davidwrighton committed:
Update docs/design/coreclr/botr/method-descriptor.md

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

davidwrighton committed:
Update src/coreclr/vm/methodtable.inl

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

davidwrighton committed:
Update src/coreclr/vm/methodtable.h

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into change_temporary_entrypoints

davidwrighton committed:
Fix GetMethodDescForSlot_NoThrow
Try removing EnsureSlotFilled
Implement IsEligibleForTieredCompilation in terms of IsEligibleForTieredCompilation_NoCheckMethodDescChunk

AaronRobinsonMSFT commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -2783,6 +2783,12 @@ TypeHandle ClassLoader::PublishType(const TypeKey *pTypeKey, TypeHandle typeHnd)
+#ifdef _DEBUG
+    if (!typeHnd.IsTypeDesc())
+    {
+        typeHnd.AsMethodTable()->GetAuxiliaryDataForWrite()->SetIsPublished();
```
Why is this only done in a `_DEBUG` build? Normally I would expect an assert or some other clearly debug-esque API being called. Instead `SetIsPublished()` sounds vaguely like a real API that should convey logic. Perhaps a comment?

davidwrighton replied:
Comment added.

davidwrighton replied:
// The IsPublished flag is used by various asserts to assure that allocations of
        // MethodTable associated memory which do not use the AllocMemTracker of the MethodTableBuilder
        // aren't permitted until the MethodTable is in a state where the MethodTable object
        // cannot be freed (except by freeing an entire LoaderAllocator)

AaronRobinsonMSFT commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -6732,13 +6720,13 @@ MethodDesc *MethodTable::MethodDataObject::GetImplMethodDesc(UINT32 slotNumber)
     {
         _ASSERTE(slotNumber < GetNumVirtuals());
-        pMDRet = m_pDeclMT->GetMethodDescForSlot(slotNumber);
+        pMDRet = m_pDeclMT->GetMethodDescForSlot_NoThrow(slotNumber);
```
The contract was updated to `THROWS`, but we're calling the "NoThrow" variant. That seems odd.

AaronRobinsonMSFT replied:
I'm still unclear on this code path. The only changes here are going from `NOTHROW` to `THROWS` and adding a method that specifically has the `NoThrow` in the name. That is unexpected. What is the "NoThrow" doing that requires the contract change?

davidwrighton replied:
Ah, this is a case where I believe I changed to a NoThrow variant after changing to a THROWS contract. I'll swap this back and verify this is true.

AaronRobinsonMSFT commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -7597,18 +7593,37 @@ Module *MethodTable::GetDefiningModuleForOpenType()
 {
     CONTRACTL {
-        NOTHROW;
+        THROWS;
```
Updated to `THROWS`, but we're calling the "NoThrow" variant seems odd.

AaronRobinsonMSFT commented on src/coreclr/vm/methodtable.h:
```diff
@@ -332,7 +332,9 @@ struct MethodTableAuxiliaryData
 
         enum_flag_HasApproxParent           = 0x0010,
-        // enum_unused                      = 0x0020,
+#ifdef _DEBUG
+        enum_flag_IsPublished               = 0x0020,
+#endif
```
What is the point of this publish logic?

davidwrighton replied:
It exists so that we can write asserts that validate that any allocation of temporary entrypoints is properly tied to the right AllocMemTracker, so that cleanup can happen correctly. Adding these asserts has shaken out a distressing number of places I missed in my manual analysis.

AaronRobinsonMSFT commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -6848,7 +6848,7 @@ VOID MethodTableBuilder::ValidateInterfaceMethodConstraints()
             // Grab the method token
             MethodTable * pMTItf = pItf->GetMethodTable();
-            CONSISTENCY_CHECK(CheckPointer(pMTItf->GetMethodDescForSlot(it.GetSlotNumber())));
+            CONSISTENCY_CHECK(CheckPointer(pMTItf->GetMethodDescForSlot_NoThrow(it.GetSlotNumber())));
```
The `STANDARD_VM_CONTRACT` permits throwing. This seems odd.

davidwrighton replied:
This is an assert. The `GetMethodDescForSlot` and `GetMethodDescForSlot_NoThrow` are designed to produce exactly the same results, given the same inputs. The only difference is in performance of repeated calls, and the fact that GetMethodDescForSlot has some side effects on the slot data structures. There is no reason to avoid the _NoThrow variant in asserts, and since the _NoThrow variant doesn't have various side effects, it is preferable to use the `GetMethodDescForSlot_NoThrow` function when in code which is only used in DEBUG code paths.

AaronRobinsonMSFT commented on src/coreclr/vm/method.cpp:
```diff
@@ -918,6 +962,95 @@ WORD MethodDesc::InterlockedUpdateFlags3(WORD wMask, BOOL fSet)
+    DWORD   dwMask = bMask;
+
+    // We need to make this operation atomic (multiple threads can play with the flags field at the same time). But the flags field
+    // is a word and we only have interlock operations over dwords. So we round down the flags field address to the nearest aligned
```
```suggestion
    // is a byte and we only have interlock operations over dwords. So we round down the flags field address to the nearest aligned
```

AaronRobinsonMSFT commented on src/coreclr/vm/method.cpp:
```diff
@@ -918,6 +962,95 @@ WORD MethodDesc::InterlockedUpdateFlags3(WORD wMask, BOOL fSet)
+
+    // We need to make this operation atomic (multiple threads can play with the flags field at the same time). But the flags field
+    // is a word and we only have interlock operations over dwords. So we round down the flags field address to the nearest aligned
+    // dword (along with the intended bitfield mask). Note that we make the assumption that the flags word is aligned itself, so we
```
```suggestion
    // dword (along with the intended bitfield mask). Note that we make the assumption that the flags byte is aligned itself, so we
```

AaronRobinsonMSFT commented on src/coreclr/vm/method.cpp:
```diff
@@ -2485,526 +2636,219 @@ MethodDesc* MethodDesc::GetMethodDescFromStubAddr(PCODE addr, BOOL fSpeculative
+            {
+                for (int i = 0; i < count; ++i)
+                {
+                    if (pMD->DetermineAndSetIsEligibleForTieredCompilation())
+                    {
+                        _ASSERTE(pMD->IsEligibleForTieredCompilation_NoCheckMethodDescChunk());
+                    }
+                    else
+                    {
+                        _ASSERTE(!pMD->IsEligibleForTieredCompilation_NoCheckMethodDescChunk());
+                    }
```
```suggestion
                    _ASSERTE(pMD->IsEligibleForTieredCompilation_NoCheckMethodDescChunk() == pMD->DetermineAndSetIsEligibleForTieredCompilation());
```

davidwrighton replied:
That is not a safe change. The `DetermineAndSetIsEligibleForTieredCompilation` method must run in release builds.

AaronRobinsonMSFT replied:
Oh, I didn't see the "AndSet". Sorry. I would then update this slightly.

```suggestion
                    bool isEligible = pMD->DetermineAndSetIsEligibleForTieredCompilation();
                    _ASSERTE(isEligible == pMD->IsEligibleForTieredCompilation_NoCheckMethodDescChunk());
```

davidwrighton committed:
Fix missing change intended in last commit

davidwrighton committed:
Fix some more IsPublished memory use issues

davidwrighton committed:
Call the right GetSlot method

davidwrighton committed:
Move another scenario to NoThrow, I think this should clear up our tests...

davidwrighton committed:
Add additional IsPublished check

davidwrighton committed:
Fix MUSL validation build error and Windows x86 build error

davidwrighton commented on src/coreclr/debug/daccess/request.cpp:
```diff
@@ -435,8 +443,15 @@ ClrDataAccess::GetMethodTableSlot(CLRDATA_ADDRESS mt, unsigned int slot, CLRDATA
+                else
+                {
+                    hr = S_OK;
+                }
```
Yes. keeping going is silly.

davidwrighton commented on src/coreclr/debug/daccess/request.cpp:
```diff
@@ -445,6 +460,88 @@ ClrDataAccess::GetMethodTableSlot(CLRDATA_ADDRESS mt, unsigned int slot, CLRDATA
+    while (slot < numVtableSlots)
+    {
+        MethodDesc* pMD = mTable->GetMethodDescForSlot_NoThrow(slot);
+        methodData.MethodDesc = HOST_CDADDR(pMD);
+        methodData.Entrypoint = mTable->GetSlot(slot);
+        methodData.DefininingMethodTable = PTR_CDADDR(pMD->GetMethodTable());
+        methodData.DefiningModule = HOST_CDADDR(pMD->GetModule());
+        methodData.Token = pMD->GetMemberDef();
```
Cleaned up.

davidwrighton commented on src/coreclr/debug/daccess/request.cpp:
```diff
@@ -445,6 +460,88 @@ ClrDataAccess::GetMethodTableSlot(CLRDATA_ADDRESS mt, unsigned int slot, CLRDATA
+        WORD slot = pMD->GetSlot();
+        if (slot >= numVtableSlots)
+        {
+            methodData.MethodDesc = HOST_CDADDR(pMD);
```
Done

davidwrighton committed:
Address code review feedback

davidwrighton committed:
Fix classcompat build

davidwrighton committed:
Update src/coreclr/vm/method.cpp

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

davidwrighton committed:
Remove assert that is invalid because TryGetMulticCallableAddrOfCode can return NULL ... and then another thread could produce a stable entrypoint and the assert could lose the race

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into change_temporary_entrypoints

AaronRobinsonMSFT commented on src/coreclr/vm/method.cpp:
```diff
@@ -2485,526 +2636,213 @@ MethodDesc* MethodDesc::GetMethodDescFromStubAddr(PCODE addr, BOOL fSpeculative
+            for (int i = 0; i < count; ++i)
+            {
+                _ASSERTE(chunkContainsEligibleMethods == pMDDebug->DetermineIsEligibleForTieredCompilationInvariantForAllMethodsInChunk());
+                pMDDebug = (MethodDesc *)(dac_cast<TADDR>(pMDDebug) + pMDDebug->SizeOf());
```
Is the `dac_cast<>` call appropriate here? This doesn't use `PTR_MethodDesc`, whcih means the call to `SizeOf()` isn't correct. It seems this should really be a `static_cast<>`.

davidwrighton replied:
static_cast can be used here instead. This code isn't compiled for the dac, so its all the same at runtime.

davidwrighton replied:
We can't actually use static_cast here, so I can use either a C style cast, reinterpret_cast, or dac_cast. Given that we commonly have logic which casts to TADDR which is required to use dac_cast, I'd like to keep the existing code, so it will work if anyone copies/pastes this code into a place where the dac actually needs to use it.

AaronRobinsonMSFT replied:
Okay. Let's keep it.

AaronRobinsonMSFT commented on src/coreclr/vm/method.cpp:
```diff
@@ -3258,7 +3095,14 @@ void MethodDesc::SetCodeEntryPoint(PCODE entryPoint)
     else if (!HasStableEntryPoint())
     {
-        SetStableEntryPointInterlocked(entryPoint);
+        if (RequiresStableEntryPoint())
+        {
+            GetOrCreatePrecode()->SetTargetInterlocked(entryPoint);
+        }
+        else
+        {
+            SetStableEntryPointInterlocked(entryPoint);
+        }
```
This code reads odd. If `RequiresStableEntryPoint()` returns `false`, then we set the stable entry point? If this is correct, I think a comment is appropriate.

davidwrighton replied:
Sigh. It is right, but yeah, this should be better structured. A RequiredStableEntryPoint is currently required have a stable entrypoint which is a Precode. Otherwise, we can generate set a stable entrypoint which isn't based on a Precode. Before this change, we would have reliably hit the HasPrecode flag above, but with making Precode generation lazy, this is a point where the Precode needs to be forced at this point. However, I think I'll fix this by changing the previous condition to be (RequiresStableEntryPoint()) and not use GetPrecode, and instead use GetOrCreatePrecode(). Then this portion of the condition can revert to be what it was before.

davidwrighton committed:
Final (hopefully) code review tweaks.

davidwrighton committed:
Its possible for GetOrCreatePrecode to be called for cases where it isn't REQUIRED. we need to handle that case.
